XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,419 @@
+by handles to perform work. Such as accepting a new connection or writing data
+to disk.
+
+
+## API
+
+### Overview
+
+```js
+// Standard way of requiring a module. Snake case follows core module
+// convention.
+const async_wrap = require('async_wrap');
+
+// List of asynchronous subsystems (e.g. CRYPTO or TCPWRAP). Can be used for
+// trace filtering of events.
+const subsystems = async_wrap.subsystems;
+
+// Retrieve a handle by its id. The id will have been passed to the init()
+// callback. Reason for this API is to prevent users from holding references
+// to the handles themselves. Thus possibly preventing GC from cleaning up
+// the handle and ever firing the destroy() callback.
+const handle = async_wrap.getHandleById(n);
+
+// init() is called during object construction. The handle will have not
+// completed construction when this callback runs. So not all fields will have
+// populated.
+function init(id, type, parentId) { }
+
+// pre() is called just before the handle's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWRAP), and should be called exactly 1
+// time for requests (e.g. FSREQWRAP).
+function pre(id) { }
+
+// post() is called just after the handle's callback has finished, and will be
+// called regardless whether the handle's callback threw. If the handle's
+// callback did throw then hasThrown will be true.
+function post(id, hasThrown) { }
+
+// destroy() is called when an AsyncWrap instance is destroyed. In cases like
+// HTTPPARSER where the resource is reused, or timers where the handle is only
+// a JS object, destroy() will be triggered manually soon after post() has
+// completed.
+function destroy(id) { }
+
+// Create a new instance and add each callback to the corresponding hook.
+var asyncHook = new AsyncHook({ init, pre, post, destroy });
+
+// Enable hooks for the current synchronous execution scope. This will ensure
+// the hooks are not in effect in case of multiple returns, or if an exception
+// is thrown.
+asyncHook.scope();
+
+// Allow callbacks of this AsyncHook instance to fire. This is not an implicit
+// action after running the constructor, and must be explicitly run to begin
+// executing callbacks.
+asyncHook.enable();
+
+// Disable listening for new asynchronous events. Though this will not prevent
+// callbacks from firing on asynchronous chains that have already run within
+// the scope of an enabled AsyncHook instance.
+asyncHook.disable();
+
+// Unlike disable(), prevent any hook callback from firing again in the future.
+// Future hooks can be added by running scope()/enable() again.
+// TODO: I am not sure all of this is possible. Some investigation is necessary.
+asyncHook.remove();
+```
+
+
+### `async_wrap`
+
+The object returned from `require('async_wrap')`.
+
+
+#### `async_wrap.subsystems`
+
+List of all subsystems that may trigger the `init` callback. These subsystems
+group relevant calls together. For example the `CRYPTO` provider encompasses
```

Comment:
use of "provider" here yet the ambiguity about subsystems remains in the next paragraph